### PR TITLE
Added support for custom ports and no. of max connections

### PR DESCRIPTION
**Pull request by FahadAhmad2001**
- Added support for both clients and servers to use ports other than 11000.
- For servers, a config.ini file is created where the server exe is located. From here you can change the port number and number of max connections.
- For clients, in the add server, if you want to use a custom port, eg 6500, just add it to the host textbox, so for example 127.0.0.1:6500. If no port is added, it will automatically use port 11000

### CODE REVIEW

**Comment by FahadAhmad2001** [comment]
#306

**Comment by FahadAhmad2001** [comment]
#306

File: NitroxClient/MonoBehaviours/Gui/MainMenu/JoinServer.cs [hunk 1]
```diff
@@ -157,7 +157,17 @@ private void StartMultiplayerClient()
             }
             catch (ClientConnectionFailedException)
             {
-                Log.InGame($"Unable to contact the remote server at: {ServerIp}:11000");
+                string Splitter = ":";
```
Comment (Measurity) [inline NitroxClient/MonoBehaviours/Gui/MainMenu/JoinServer.cs]: Instead of using an if check on the ipaddress string I would just always include the port number given. You can use `multiplayerSession.Port` if you added it.

`Log.InGame($"Unable to contact the remote server at: {multiplayerSession.IpAddress}:{multiplayerSession.Port}");`

File: NitroxServer/Communication/UdpServer.cs [hunk 2]
```diff
@@ -156,8 +156,34 @@ private Connection GetConnection(long remoteIdentifier)
         private NetPeerConfiguration BuildNetworkConfig()
         {
             NetPeerConfiguration config = new NetPeerConfiguration("Nitrox");
-            config.Port = 11000;
-            config.MaximumConnections = 100;
+            string INIContents;
```
Comment (Measurity) [inline NitroxServer/Communication/UdpServer.cs]: You can use `File.ReadAllText(@".\config.ini")` instead of `StreamReader.ReadToEnd`. Saves 2 lines.

The UdpServer is getting messy if it also contains INI reading and parsing logic. It would be better to put an INI file reading/writing/parsing class in the `NitroxModel` project and then use it here. Though I'm ok with it to keep it like this for now because of the work required.

File: NitroxServer/Communication/UdpServer.cs [hunk 3]
```diff
@@ -156,8 +156,34 @@ private Connection GetConnection(long remoteIdentifier)
...
+                ParseINIContents=SplitINIContents[0].Split(equals);
+                PortNumber = int.Parse(ParseINIContents[1]);
+                ParseINIContents = SplitINIContents[1].Split(equals);
+                MaxConn = int.Parse(ParseINIContents[1]);
+            }
+            else
+            {
+                string NewINIContents;
+                NewINIContents = "[NetworkSettings]\nDefaultPortNumber=11000;\nMaxConnections=100;";
+                StreamWriter WriteINI = new StreamWriter(@".\config.ini");
```
Comment (Measurity) [inline NitroxServer/Communication/UdpServer.cs]: Similar to code above you can use `File.WriteAllText(@".\config.ini", NewINIContents);` instead of StreamWriter.

File: NitroxClient/Communication/UdpClient.cs [hunk 4]
```diff
@@ -26,13 +26,28 @@ public UdpClient(DeferringPacketReceiver packetManager)
         public void Start(string ipAddress)
```
[the remark this answers is not in the document]
Comment (FahadAhmad2001) [reply NitroxClient/Communication/UdpClient.cs]: As this causes errors with NitroxTest and ClientTester as port number is not specified in these projects, so in these 2 projects, do I set the port number as 11000, like this:

File: NitroxServer/Communication/UdpServer.cs (hunk 2 above)
Comment (FahadAhmad2001) [reply NitroxServer/Communication/UdpServer.cs]: Fixed this in the new commit

**Comment by FahadAhmad2001** [comment]
Added the requested changes in the new commit

File: NitroxClient/Communication/Abstract/IClient.cs [hunk 5]
```diff
@@ -8,7 +8,7 @@ namespace NitroxClient.Communication.Abstract
     public interface IClient
     {
         bool IsConnected { get; }
-        void Start(string ipAddress);
+        void Start(string ipAddress, int ServerPort);
```
Comment (Measurity) [inline NitroxClient/Communication/Abstract/IClient.cs]: Method parameters should be camcelCase. `serverPort`

Comment (FahadAhmad2001) [reply NitroxClient/Communication/Abstract/IClient.cs]: Should I change it to serverPort

Comment (Measurity) [reply NitroxClient/Communication/Abstract/IClient.cs]: Yes, that's the naming convention for method parameters.

File: NitroxServer/Communication/UdpServer.cs [hunk 6]
```diff
@@ -156,8 +156,30 @@ private Connection GetConnection(long remoteIdentifier)
         private NetPeerConfiguration BuildNetworkConfig()
         {
             NetPeerConfiguration config = new NetPeerConfiguration("Nitrox");
-            config.Port = 11000;
-            config.MaximumConnections = 100;
+            string INIContents;
+            int PortNumber, MaxConn;
+            if (File.Exists(@".\config.ini"))
```
[the remark this answers is not in the document]
Comment (FahadAhmad2001) [reply NitroxServer/Communication/UdpServer.cs]: I created a new class to read the config, but do I call it before server.Start(), or call it when config.Port and config.MaximumConnections are being set

[the remark this answers is not in the document]
Comment (Measurity) [reply NitroxServer/Communication/UdpServer.cs]: Ideally, we should pass a custom config class into the UdpServer constructor so that file reading happens before the `UdpServer` is created.

Something like: `new UdpServer(..., UdpServerConfig.FromFile("config.ini"));` and then in the `FromFile` function you use the IniReader/Writer class you created and store the relevant INI fields in the custom config class.

Lastly, map the custom config class to `NetPeerConfiguration` inside UdpServer constructor so that Lidgren classes (like `NetPeerConfiguration`) don't leave the `UdpServer` class.

To answer your question: [`Server.Start()`](https://github.com/SubnauticaNitrox/Nitrox/blob/4cececc388920cef22b0f83313b25ff01ee0a939/NitroxServer/Server.cs#L44) sounds like the best location because reading files is slow. Move the `new UdpServer()` from the `Server` class constructor into the `Server.Start()` to do this.

[the remark this answers is not in the document]
Comment (Sunrunner37) [reply NitroxServer/Communication/UdpServer.cs]: I'm not sure if the config class should be specific to UdpServer.  I think that we will have a wide range of server configurations that we'll want to access in the future. (An example being how often we auto save).  When the server first starts we can parse the config file and pass around a data structure that encapsulates all server configs.  A subset of these configs could be NetworkConfigs.

[the remark this answers is not in the document]
Comment (Measurity) [reply NitroxServer/Communication/UdpServer.cs]: You're right. A more abstract server config file would be better. But I'm not sure how much time  @FahadAhmad2001 has to implement all this. Please let us know what you think.

[the remark this answers is not in the document]
Comment (FahadAhmad2001) [reply NitroxServer/Communication/UdpServer.cs]: I tried moving the new UdpServer() to server.Start(), but then I have to change it from private readonly UdpServer to private UdpServer. Also, I cant make Server.Start() static because then EnablePeriodicSaving() does not work (unless I make that static as well).

So, should I read from the config file before UdpServer is created, ie in Server.cs

This way, values such as time before auto save can also be read and set here
The code in UdpServer.cs:

and

Also, here is the ServerConfigReader class:

Finally, should I place the serverConfigreader class in NitroxModel or in NitroxServer?

[the remark this answers is not in the document]
Comment (Measurity) [reply NitroxServer/Communication/UdpServer.cs]: Looks good so far. Yes, the config should be read before the `UdpServer` and `Server` are started so that both can be configurated by the same INI file. It would be better to read the config file and pass the config class instance to the `Server` constructor instead of the file path itself. This way you don't have to bother making things static because the IO reading happens before the server constructor. I'll let you decide whether to keep `ServerConfigReader` static or pass it as an instance to the `Server` and `UdpServer` constructors. Both work.

So in short:
- Move the ServerConfigReader.ReadServerConfig(configPath) to `Program` class where the `Server` instance is created.
- Then read out the `ServerConfigReader` class in `Server` (if you have any server options in the INI file that is) and `UdpServer` (get the port from INI file).

Discard what I said about making `UdpServer.Start()` static in the comments above.

**Comment by FahadAhmad2001** [comment]
Should I add/change anything?

File: NitroxClient/Communication/Abstract/IMultiplayerSessionState.cs [hunk 7]
```diff
@@ -7,6 +7,7 @@ public interface IMultiplayerSessionState
     {
         IClient Client { get; }
         string IpAddress { get; }
+        int serverPort { get; }
```
Comment (Measurity) [inline NitroxClient/Communication/Abstract/IMultiplayerSessionState.cs]: Could you make `serverPort` pascal case as `ServerPort`? Only for this property. We use pascal case for properties.

File: NitroxClient/Communication/MultiplayerSession/MultiplayerSessionManager.cs [hunk 8]
```diff
@@ -28,16 +28,18 @@ internal MultiplayerSessionManager(IClient client, IMultiplayerSessionConnection
 
         public IClient Client { get; }
         public string IpAddress { get; private set; }
+        public int serverPort{ get; private set; }
```
Comment (Measurity) [inline NitroxClient/Communication/MultiplayerSession/MultiplayerSessionManager.cs]: Same case here, also pascal case for `ServerPort`

File: NitroxClient/MonoBehaviours/Gui/MainMenu/JoinServer.cs [hunk 9]
```diff
@@ -153,11 +153,11 @@ private void StartMultiplayerClient()
 
             try
             {
-                multiplayerSession.Connect(ServerIp);
+                multiplayerSession.Connect(ServerIp,serverPort);
             }
             catch (ClientConnectionFailedException)
             {
-                Log.InGame($"Unable to contact the remote server at: {ServerIp}:11000");
+                Log.InGame($"Unable to contact the remote server at: {ServerIp}:{serverPort.ToString()}");
```
Comment (Measurity) [inline NitroxClient/MonoBehaviours/Gui/MainMenu/JoinServer.cs]: `ToString()` is not needed when you use interpolated strings (or string.Format).

File: NitroxServer/Server.cs [hunk 10]
```diff
@@ -13,21 +15,22 @@ public class Server
...
+        private ServerConfigReader ServerOptions;
         public static Server Instance;
 
         public void Save()
         {
             worldPersistence.Save(world);
         }
 
-        public Server()
+        public Server(ServerConfigReader ConfigReader)
```
Comment (Measurity) [inline NitroxServer/Server.cs]: parameters for methods/functions are camelCase. `ConfigReader` to `configReader`.
